### PR TITLE
Plausibility model

### PR DESCRIPTION
**Pull request by ringgaard**
The plausibility model is used for predicting the probability of a fact being true given a known set of fact for an item. The plausibility model is a simple two-layer feed-forward model with embeddings for the premise and hypothesis facts. The facts come from the fact lexicon, and fact are now grouped, so each group consists of a base fact together with the fact derived from the base fact. The internal representation of facts have been changed so you don't need a writable store to extract facts. I have also implemented a Python interface for the plausibility model that can score a property/value pair for an item.

### CODE REVIEW

File: sling/nlp/embedding/fact-plausibility.cc [hunk 1]
```diff
@@ -0,0 +1,547 @@
...
+            num_training_examples_->Increment();
+          }
+        }
+      }
+
+      // Update parameters.
+      optimizer_mu_.Lock();
+      optimizer_->Apply(gradients);
+      loss_sum_ += epoch_loss;
+      loss_count_ += batches_per_update_ * batch_size_ * batch_size_;
```
Comment (rahul1980) [inline sling/nlp/embedding/fact-plausibility.cc]: Shouldn't we omit counting those cases where the hypothesis of a -ve example was in the premise?

File: sling/nlp/embedding/fact-plausibility.cc [hunk 2]
```diff
@@ -0,0 +1,547 @@
...
+      } else {
+        // Add fact group to premise.
+        for (int f = start; f < end; ++f) {
+          premise->push_back(facts.get(f).AsInt());
+        }
+      }
+    }
+
+    if (premise->size() >= max_features_) {
+      premise->resize(max_features_ - 1);
```
Comment (rahul1980) [inline sling/nlp/embedding/fact-plausibility.cc]: why -1? Is it so that we have space to pad the final feature vector with a sentinel -1 at the end? I thought the sentinel was only required for embedding lookup if we were not specifying all max_features entries.

File: sling/nlp/embedding/fact-plausibility.cc [hunk 3]
```diff
@@ -0,0 +1,547 @@
...
+
+        // Do forward and backward propagation for each premise/hypothesis pair.
+        // Each sampled item is a positive example. Negative examples are
+        // generated by using the premise from one item and the hypothsis from
+        // another item.
+        for (int i = 0; i < batch_size_; ++i) {
+          for (int j = 0; j < batch_size_; ++j) {
+            // Check that main hypothesis is not in premise for negative
+            // examples.
+            if (i != j && Has(premises[i], hypotheses[j][0])) {
```
Comment (rahul1980) [inline sling/nlp/embedding/fact-plausibility.cc]: Is it correct to say that 'main hypothesis', i.e. the one at index 0, is the most fine-grained (i.e. not expanded) version of the fact?

File: sling/nlp/embedding/fact-plausibility.cc [hunk 4]
```diff
@@ -0,0 +1,547 @@
...
+        int num_groups = instance.Get(p_groups_).AsArray().length();
+        int heldout = rnd.UniformInt(num_groups);
+        Split(instance, &premises[i], &hypotheses[i], heldout);
+      }
+
+      // Compute scores for all pairs in batch.
+      for (int i = 0; i < batch_size; ++i) {
+        for (int j = 0; j < batch_size; ++j) {
+          // Check that main hypothesis is not in premise for negative examples.
+          if (i != j && Has(premises[i], hypotheses[j][0])) continue;
```
Comment (rahul1980) [inline sling/nlp/embedding/fact-plausibility.cc]: Split() truncates 'premises' and 'hypotheses' to max_features - 1, so Has() here can erroneously return false, since it could be looking for the hypothesis in a truncated list of premise facts. This may or may not be a big issue, but it is straightforward to change Split() not to do any truncation, so I would suggesting doing that. 
The truncation can be delegated to Copy().

File: python/task/embedding.py [hunk 5]
```diff
@@ -154,19 +154,55 @@ def train_fact_embeddings(self):
     with self.wf.namespace("fact-embeddings"):
       trainer = self.wf.task("fact-embeddings-trainer")
       trainer.add_params({
-        "batch_size": 1024,
-        "embedding_dims": 128,
+        "batch_size": 256,
```
[the remark this answers is not in the document]
Comment (ringgaard) [reply python/task/embedding.py]: I tried to see if I could improve the (percisved) quality of the fact embeddings, so they looked a little better with these hyper parameters. However, I ended up not using them for the plausibility model.

File: sling/frame/store.h [hunk 6]
```diff
@@ -1010,6 +1010,7 @@ class Store {
   // become dependent on the store.
   uint64 Fingerprint(Handle handle, bool byref = false, uint64 seed = 0) const;
   uint64 Fingerprint(ArrayDatum *array, int begin, int end, int step) const;
```
[the remark this answers is not in the document]
Comment (ringgaard) [reply sling/frame/store.h]: I think this might be a bug. Shouldn't we always compute list fingerprints by value?

File: sling/nlp/embedding/BUILD [hunk 7]
```diff
@@ -65,3 +65,39 @@ cc_library(
...
 
+cc_library(
+  name = "fact-plausibility",
+  srcs = ["fact-plausibility.cc"],
+  deps = [
+    "//sling/base",
+    "//sling/file:textmap",
+    "//sling/frame:object",
+    "//sling/frame:store",
+    "//sling/frame:serialization",
```
[the remark this answers is not in the document]
Comment (ringgaard) [reply sling/nlp/embedding/BUILD]: Done

File: sling/nlp/embedding/BUILD [hunk 8]
```diff
@@ -65,3 +65,39 @@ cc_library(
...
+
+cc_library(
+  name = "plausibility-model",
+  srcs = ["plausibility-model.cc"],
+  hdrs = ["plausibility-model.h"],
+  deps = [
+    "//sling/base",
+    "//sling/frame:object",
+    "//sling/frame:store",
+    "//sling/frame:serialization",
```
[the remark this answers is not in the document]
Comment (ringgaard) [reply sling/nlp/embedding/BUILD]: Done

File: sling/nlp/embedding/fact-embeddings.cc [hunk 9]
```diff
@@ -44,8 +44,8 @@ class FactLexiconExtractor : public Process {
     // Get parameters.
     int64 bloom_size = task->Get("bloom_size", 4000000000L);
     int bloom_hashes = task->Get("bloom_hashes", 4);
-    int fact_threshold = task->Get("fact_threshold", 10);
-    int category_threshold = task->Get("category_threshold", 10);
+    int fact_threshold = task->Get("fact_threshold", 30);
+    int category_threshold = task->Get("category_threshold", 30);
```
[the remark this answers is not in the document]
Comment (ringgaard) [reply sling/nlp/embedding/fact-embeddings.cc]: We produced low-quality embeddings for the small fact and categories. This removes ~2/3 of the lexicon entries.

File: sling/nlp/embedding/fact-plausibility.cc [hunk 10]
```diff
@@ -0,0 +1,547 @@
...
+using namespace myelin;
+
+// Fact plausibility flow.
+struct FactPlausibilityFlow : public Flow {
+  // Build fact plausibility model.
+  void Build(const Transformations &library, bool learn) {
+    scorer = AddFunction("scorer");
+    FlowBuilder f(this, scorer);
+    auto *embeddings =
+        f.Random(f.Parameter("embeddings", DT_FLOAT, {facts, dims}));
```
[the remark this answers is not in the document]
Comment (ringgaard) [reply sling/nlp/embedding/fact-plausibility.cc]: This is also what it is set to in FactPlausibilityTrainer::Build.

File: sling/nlp/embedding/fact-plausibility.cc [hunk 11]
```diff
@@ -0,0 +1,547 @@
...
+
+          // Split instance into premise and hypothesis.
+          int num_groups = instance.Get(p_groups_).AsArray().length();
+          int heldout = rnd.UniformInt(num_groups);
+          Split(instance, &premises[i], &hypotheses[i], heldout);
+        }
+
+        // Do forward and backward propagation for each premise/hypothesis pair.
+        // Each sampled item is a positive example. Negative examples are
+        // generated by using the premise from one item and the hypothsis from
```
[the remark this answers is not in the document]
Comment (ringgaard) [reply sling/nlp/embedding/fact-plausibility.cc]: Fixed.

File: sling/nlp/embedding/fact-plausibility.cc (hunk 3 above)
Comment (ringgaard) [reply sling/nlp/embedding/fact-plausibility.cc]: Yes, the base fact is the first in the group.

File: sling/nlp/embedding/fact-plausibility.cc [hunk 12]
```diff
@@ -0,0 +1,547 @@
...
+
+            // Set premise and hypothesis for example.
+            Copy(premises[i], premise);
+            Copy(hypotheses[j], hypothesis);
+
+            // Compute plausibility scores.
+            scorer.Compute();
+
+            // Compute accuracy.
+            batch_benchmark.add(i == j, logits[1] > logits[0]);
```
[the remark this answers is not in the document]
Comment (ringgaard) [reply sling/nlp/embedding/fact-plausibility.cc]: Plausibility is a score between 0 and 1, so it would not be correct to say that we map '0' to implausible and '1' to plausible. We train by giving true facts a score of 1 and (presumably) false facts a score of 0. I added a comment about the order of the logits.

File: sling/nlp/embedding/fact-plausibility.cc (hunk 1 above)
Comment (ringgaard) [reply sling/nlp/embedding/fact-plausibility.cc]: Yes, otherwise the avg loss will be slightly off. I have fixed this.

File: sling/nlp/embedding/fact-plausibility.cc (hunk 4 above)
Comment (ringgaard) [reply sling/nlp/embedding/fact-plausibility.cc]: I thought it was easier to ensure that we newer overflowed the feature vector so I don't have to check for the case where there is no room for the terminating -1. You're right that we would miss a contradiction for a truncated item, but we have very few overflows, so I am not sure this matters.

File: sling/nlp/embedding/fact-plausibility.cc (hunk 2 above)
Comment (ringgaard) [reply sling/nlp/embedding/fact-plausibility.cc]: As stated above, we always leave room for the terminating -1 in the feature vector.

File: sling/nlp/kb/facts.h [hunk 13]
```diff
@@ -180,24 +187,60 @@ class Facts {
...
+
+  // Return interval for fact, i.e. fact(i) = list[begin(i):end(i)].
+  int begin(int i) const { return i == 0 ? 0 : delimiters_[i - 1]; }
+  int end(int i) const { return delimiters_[i]; }
+
+  // Return base property for fact, i.e. first value in fact path.
+  Handle first(int i) const { return list_[begin(i)]; }
+
+  // Return fact value, i.e. last value in fact path.
+  Handle last(int i) const { return list_[end(i) - 1]; }
```
[the remark this answers is not in the document]
Comment (ringgaard) [reply sling/nlp/kb/facts.h]: I must admit that I quite like the begin/end, first/last naming. Calling it property would be misleading.

File: sling/task/learner.cc [hunk 14]
```diff
@@ -56,7 +57,16 @@ void LearnerTask::Train(Task *task, myelin::Network *model) {
...
-    Evaluate(epoch_, model);
+    if (!Evaluate(epoch_, model)) {
+      done_ = true;
+      break;
+    }
+
+    // Checkpoint model.
+    if (epoch_ >= last_checkpoint_ + checkpoint_interval_) {
+      Checkpoint(epoch_, model);
+      last_checkpoint_ += checkpoint_interval_;
```
[the remark this answers is not in the document]
Comment (ringgaard) [reply sling/task/learner.cc]: The epoch_ variable is an atomic counter, which can be updated while running, so I think it is safer to do it this way to avoid drift in the checkpoint intervals.
